src/oscar/apps/dashboard/vouchers/forms.py

```diff
@@ -1,99 +1,85 @@
-            pass
-        else:
-            if (not self.voucher) or (voucher.id != self.voucher.id):
-                raise forms.ValidationError(_("The name '%s' is already in"
-                                              " use") % name)
-        return name
+    class Meta:
+        model = Voucher
+        fields = [
+            'name',
+            'code',
+            'start_datetime',
+            'end_datetime',
+            'usage',
+        ]
+        widgets = {
+            'start_datetime': widgets.DateTimePickerInput(),
+            'end_datetime': widgets.DateTimePickerInput(),
+        }
 
     def clean_code(self):
-        code = self.cleaned_data['code'].strip().upper()
-        if not code:
-            raise forms.ValidationError(_("Please enter a voucher code"))
-        try:
-            voucher = Voucher.objects.get(code=code)
-        except Voucher.DoesNotExist:
-            pass
-        else:
-            if (not self.voucher) or (voucher.id != self.voucher.id):
-                raise forms.ValidationError(_("The code '%s' is already in"
-                                              " use") % code)
-        return code
-
-    def clean(self):
-        cleaned_data = super().clean()
-        start_datetime = cleaned_data.get('start_datetime')
-        end_datetime = cleaned_data.get('end_datetime')
-        if start_datetime and end_datetime and end_datetime < start_datetime:
-            raise forms.ValidationError(_("The start date must be before the"
-                                          " end date"))
-        return cleaned_data
+        return self.cleaned_data['code'].strip().upper()
 
 
 class VoucherSearchForm(forms.Form):
     name = forms.CharField(required=False, label=_("Name"))
     code = forms.CharField(required=False, label=_("Code"))
-    is_active = forms.BooleanField(required=False, label=_("Is Active?"))
-    in_set = forms.BooleanField(
-        required=False, label=_("In Voucherset?"))
+    offer_name = forms.CharField(required=False, label=_("Offer name"))
+    is_active = forms.NullBooleanField(required=False, label=_("Is active?"))
+    in_set = forms.NullBooleanField(required=False, label=_("In voucher set?"))
+    has_offers = forms.NullBooleanField(required=False, label=_("Has offers?"))
+
+    basic_fields = [
+        'name',
+        'code',
+        'is_active',
+        'in_set',
+    ]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Give the "unknown" choice a more meaningful label than the default of "Unknown"
+        NULL_BOOLEAN_FIELD_CHOICES = (
+            ('unknown', _('---------')),
+            ('true', _('Yes')),
+            ('false', _('No')),
+        )
+        self.fields['is_active'].widget.choices = NULL_BOOLEAN_FIELD_CHOICES
+        self.fields['in_set'].widget.choices = NULL_BOOLEAN_FIELD_CHOICES
+        self.fields['has_offers'].widget.choices = NULL_BOOLEAN_FIELD_CHOICES
```

Comment:
If you could specify this in the form field definition that would be preferrable. Since I'm seeing the need for this so many times, maybe just create a special widget class that does what you need? And specify it to be used in the form field.